Sony IMX290/IMX327 support

6by9 opened the pull request:
Sony IMX290 and IMX327 are broadly compatible sensors.
Backport the mainline driver and latest fixes (in linuxtv/master, not yet in torvalds linux/master).
Adds additional controls, alternate oscillator frequency support, and an overlay to load the thing.

I still need to compile test this branch, and make sure that the changes in the header comments to the driver are in a sensible patch (I think they may have got mixed up with other stuff).

Mani-Sadhasivam committed:
media: dt-bindings: media: i2c: Add IMX290 CMOS sensor binding

Commit 8a97a4676f8b1badcd9cfbed2b081342847bb1b1 upstream.

Add devicetree binding for IMX290 CMOS image sensor. Let's also
add MAINTAINERS entry for the binding and driver.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Reviewed-by: Rob Herring <robh@kernel.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

Mani-Sadhasivam committed:
media: i2c: Add IMX290 CMOS image sensor driver

Commit 828dbc299278065b634e913d2700d254a3224853 upstream.

Add driver for Sony IMX290 CMOS image sensor driver. The driver only
supports I2C interface for programming and MIPI CSI-2 for sensor output.

[Sakari Ailus: Rewrapped a few lines over 80 chars a little.]

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+samsung@kernel.org>

6by9 committed:
dtoverlays: Add an overlay for the Sony IMX290 image sensor

Adds an overlay to configure the IMX290 image sensor.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: imx290: Correct range for V4L2_CID_GAIN to 0-238

The datasheet lists the gain as being 0.0 to 72.0dB in 0.3dB steps, which
makes 238 steps total.
Correct the 0-72 range defined in the driver.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: imx290: Convert HMAX setting into V4L2_CID_HBLANK

Userspace needs to know HBLANK if it is to work out exposure times
and frame rates, therefore convert it to map onto V4L2_CID_HBLANK

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: imx290: Add support for V4L2_CID_VBLANK

In order to calculate framerate and durations userspace needs
the vertical blanking information. This can be configurable,
and indeed the datasheet lists different values for VBLANK for
the 1080p and 720p modes.

Add the new control, and adopt the datasheet values for each mode.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: imx290: Add exposure control to the driver.

Adds support for V4L2_CID_EXPOSURE so that userspace can control
the sensor exposure time.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: imx290: Add H and V flip controls

The sensor supports horizontal and vertical flips, so support them
through V4L2_CID_HFLIP and V4L2_CID_VFLIP.

This sensor does NOT change the Bayer order when changing the
direction of readout, therefore no special handling is required for
that.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

andrey-konovalov committed:
media: i2c: imx290: set the format before VIDIOC_SUBDEV_G_FMT is called

Commit d46cfdc86c30d5ec768924f0b1e2683c8d20b671 upstream.

With the current driver 'media-ctl -p' issued right after the imx290 driver
is loaded prints:
pad0: Source
             [fmt:unknown/0x0]

The format value of zero is due to the current_format field of the imx290
struct not being initialized yet.

As imx290_entity_init_cfg() calls imx290_set_fmt(), the current_mode field
is also initialized, so the line which set current_mode to a default value
in driver's probe() function is no longer needed.

Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Reviewed-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

andrey-konovalov committed:
media: i2c: imx290: fix the order of the args in SET_RUNTIME_PM_OPS()

Commit 8d2d1bedb1b9af3e0c039a4444858da7b6da71f8 upstream.

The macro is defined as SET_RUNTIME_PM_OPS(suspend_fn, resume_fn, idle_fn),
so imx290_power_off must be the 1st arg, and imx290_power_on the 2nd.

Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Reviewed-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

andrey-konovalov committed:
media: i2c: imx290: fix reset GPIO pin handling

Commit 3909a92d7df622b41b9ceeeea694e641cad7667b upstream.

According to https://www.kernel.org/doc/Documentation/gpio/consumer.txt,

- all of the gpiod_set_value_xxx() functions operate with the *logical*
value. So in imx290_power_on() the reset signal should be cleared
(de-asserted) with gpiod_set_value_cansleep(imx290->rst_gpio, 0), and in
imx290_power_off() the value of 1 must be used to apply/assert the reset
to the sensor. In the device tree the reset pin is described as
GPIO_ACTIVE_LOW, and gpiod_set_value_xxx() functions take this into
account,

- when devm_gpiod_get_optional() is called with GPIOD_ASIS, the GPIO is
not initialized, and the direction must be set later; using a GPIO
without setting its direction first is illegal and will result in undefined
behavior. Fix this by using GPIOD_OUT_HIGH instead of GPIOD_ASIS (this
asserts the reset signal to the sensor initially).

Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Reviewed-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

Mani-Sadhasivam committed:
media: i2c: imx290: Add support for 2 data lanes

Commit 97589ad61c730e0f486635c6c19fa25ab8e8f29d upstream.

The IMX290 sensor can output frames with 2/4 CSI2 data lanes. This commit
adds support for 2 lane mode in addition to the 4 lane and also
configuring the data lane settings in the driver based on system
configuration.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

Mani-Sadhasivam committed:
media: i2c: imx290: Add configurable link frequency and pixel rate

Commit 98e0500eadb772e1be32d8e369fcc3b7bcac93ed upstream.

IMX290 operates with multiple link frequency and pixel rate combinations.
The initial driver used a single setting for both but since we now have
the lane count support in place, let's add configurable link frequency
and pixel rate.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

Mani-Sadhasivam committed:
media: i2c: imx290: Add support for test pattern generation

Commit a58df1f9e4885eaf3d0663574a217e513821a9f0 upstream.

Add support for generating following test patterns by IMX290:

* Sequence Pattern 1
* Horizontal Color-bar Chart
* Vertical Color-bar Chart
* Sequence Pattern 2
* Gradation Pattern 1
* Gradation Pattern 2
* 000/555h Toggle Pattern

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

Mani-Sadhasivam committed:
media: i2c: imx290: Add RAW12 mode support

Commit c566ac01ceaa02450acc155201772c0623530e76 upstream.

IMX290 is capable of outputting frames in both Raw Bayer (packed) 10 and
12 bit formats. Since the driver already supports RAW10 mode, let's add
the missing RAW12 mode as well.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

Mani-Sadhasivam committed:
media: i2c: imx290: Add support to enumerate all frame sizes

Commit 3b867fb641d884b714fba390ae866714ba475f29 upstream.

Add support to enumerate all frame sizes supported by IMX290. This is
required for using with userspace tools such as libcamera.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

Mani-Sadhasivam committed:
media: i2c: imx290: Move the settle time delay out of loop

Commit 6544af9b04b4484867c234ba0be1b5008e4a14ee upstream.

The 10ms settle time is needed only at the end of all consecutive
register writes. So move the delay to outside of the for loop of
imx290_set_register_array().

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

andrey-konovalov committed:
media: i2c: imx290: set bus_type before calling v4l2_fwnode_endpoint_alloc_parse()

Commit a270675875829b6d46eb9e38960fd6019555ebb8 upstream.

The bus_type field of v4l2_fwnode_endpoint structure passed as the argument
to v4l2_fwnode_endpoint_alloc_parse() function must be initiaized.
Set it to V4L2_MBUS_CSI2_DPHY, and check for -ENXIO which is returned
when the requested media bus type doesn't match the fwnode.

Also remove v4l2_fwnode_endpoint field from struct imx290 as it is only
needed in the probe function: use the local variable for this purpose.

Signed-off-by: Andrey Konovalov <andrey.konovalov@linaro.org>
Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab+huawei@kernel.org>

6by9 committed:
media: i2c: imx290: Add support for 74.25MHz clock

The existing driver only supported a clock of 37.125MHz, but the
sensor also supports 74.25MHz.

Add the relevant register modifications to support this alternate
clock frequency.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: dt-bindings: media: i2c: Add mono version to IMX290 bindings

The IMX290 module is available as either monochrome or colour and
the variant is not detectable at runtime.

Add a new compatible string for the monochrome version.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media : i2c: imx290: Add support for the mono sensor variant.

The IMX290 module is available as either mono or colour (Bayer).

Update the driver so that it can advertise the correct mono
formats instead of the colour ones.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
defconfig: Add Sony IMX290 image sensor driver to all Pi defconfigs

Adds the mainline IMX290 sensor driver (with extra features) to
the default configs.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Rebased and removed the WIP tag.
I think this is good for review now.

6by9 commented:
@andrey-konovalov FYI Your/Mani's patches backport, and my couple of extras on top.

Hopefully @davidplowman will be getting it running with libcamera next week when his module arrives.

pelwell commented:
Is there a preferred merge order between this and #3688?

6by9 commented:
> Is there a preferred merge order between this and #3688?

There shouldn't be. This should work with the older Unicam driver. The imx290 dtoverlay may have a couple of extra bits that the older unicam driver doesn't care about, but those will be ignored.

6by9 commented:
Oh, and the reason I've split the dtoverlay into a imx290 and imx290_327 overlay is so that we can add an imx327 overlay easily if we get to a point where we support the differences between the two (120fps mode on IMX290).

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -1430,6 +1430,21 @@ Load:   dtoverlay=imx219
+        variants. NB This currently uses 4 CSI2 data lanes and therefore will
+        only work on a CM.
+Load:   dtoverlay=imx290,<param>
+Params: 4lanes              Enable 4 CSI2 lanes. This requires a Compute Module
```
The overlay uses the name `4lane` (singular). And (annoyingly, I know) parameter descriptions _must_ be indented to column 33, i.e. with 32 leading spaces (this is column 29). e.g.:
```
Params: 4lane                   Enable 4 CSI2 lanes. This requires a Compute
                                Module (1, 3, or 4).
        clock-frequency         Sets the clock frequency to match that used on
                                the board. Modules from Vision Components use
                                37.125MHz (the default), whilst those from
                                Innomaker use 74.25MHz.
        mono                    Denote that the module is a mono sensor.
```

6by9 replied:
Done

pelwell commented on arch/arm/boot/dts/overlays/imx290_327-overlay.dts:
```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: GPL-2.0-only
```
Could you make this a `.dtsi` file instead to make it clear to the reader (and the checker) that it isn't a complete overlay that needs documentation and a Makefile entry?

6by9 replied:
Done

pelwell commented on drivers/media/i2c/imx290.c:
```diff
@@ -40,6 +40,9 @@ enum imx290_clk_index {
 #define IMX290_HMAX_LOW 0x301c
 #define IMX290_HMAX_HIGH 0x301d
+#define IMX290_HMAX_MIN_2LANE 4400 /* Min of 4400 pixels = 30fps */
+#define IMX290_HMAX_MIN_4LANE 2200 /* Min of 200 pixels = 60fps */
```
Typo - `200` should be `2200`.

6by9 replied:
Done

pelwell commented on drivers/media/i2c/imx290.c:
```diff
@@ -518,6 +523,26 @@ static int imx290_set_gain(struct imx290 *imx290, u32 value)
+	u32 hmax = val + imx290->current_mode->width;
+	int ret;
+
+	ret = imx290_write_reg(imx290, IMX290_HMAX_LOW, (hmax & 0xff));
```
Why are IMX290_HMAX_LOW & _HIGH written explicitly as two 8-bit registers, rather than using imx290_write_buffered_reg(imx290, IMX290_HMAX_LOW, 2, hmax)? Have I missed a subtlety?

6by9 replied:
I'd meant to query this on the mailing list, but missed my chance as they got merged just before I sent my comment and Tested-by: tag.

6by9 replied:
Not that my diff here is only moving the function from below as it needs to be called from imx290_set_ctrl

pelwell replied:
OK - it was just an observation. I initially had a comment against the IMX290_VMAX support before realising that it was correct.

6by9 replied:
I've added an extra patch that converts imx290_set_hmax to use imx290_write_buffered_reg

pelwell commented on drivers/media/i2c/imx290.c:
```diff
@@ -618,6 +622,16 @@ static int imx290_set_ctrl(struct v4l2_ctrl *ctrl)
+		if (ret)
+			break;
+		val &= ~0x03;
+		val |= imx290->vflip->val | (imx290->hflip->val << 1);
```
Does V4L2 guarantee these are always 0 or 1?

6by9 replied:
Yes.
https://elixir.bootlin.com/linux/latest/source/drivers/media/v4l2-core/v4l2-ctrls.c#L1153
Type V4L2_CTRL_TYPE_BOOLEAN, min 0, max 1

pelwell commented on drivers/media/i2c/imx290.c:
```diff
@@ -648,7 +648,7 @@ static int imx290_power_off(struct device *dev)
 }
```
Typo `uptream` in commit message.

6by9 replied:
Done.

pelwell commented on drivers/media/i2c/imx290.c:
```diff
@@ -312,6 +358,11 @@ static const struct imx290_mode imx290_modes_2lanes[] = {
 		.data = imx290_1080p_settings,
 		.data_size = ARRAY_SIZE(imx290_1080p_settings),
+		.clk_data = {
+			imx290_37_125mhz_clock_1080p,
```
`[CLK_37_125] = ...`? Or a comment about the order? I'll take a no.

6by9 replied:
There is a comment in the definition of struct imx290_mode
```
	/* Clock setup can vary. Index as enum imx290_clk_index */
	const struct imx290_regval *clk_data[2];
```
but it doesn't hurt to be explicit.
Done.

pelwell commented on drivers/media/i2c/imx290.c:
```diff
@@ -1091,6 +1106,12 @@ static s64 imx290_check_link_freqs(const struct imx290 *imx290,
+static const struct of_device_id imx290_of_match[] = {
+	{ .compatible = "sony,imx290", .data = (void *)0 },
+	{ .compatible = "sony,imx290-mono", .data = (void *)1 },
+	{ /* sentinel */ }
```
Oh look - a initialising a structure to zero without the aid of a zero.

6by9 replied:
:-P

pelwell commented on drivers/media/i2c/imx290.c:
```diff
@@ -1115,6 +1137,17 @@ static int imx290_probe(struct i2c_client *client)
+	match = of_match_device(imx290_of_match, dev);
+	if (!match)
+		return -ENODEV;
+	if (!match->data) {
```
It's shame you can't just use the match data directly, but those ARRAY_SIZE()s make it hard to find a neat way to do it.

6by9 replied:
There are only ever two formats, so the arrays should always be the same size.

If I can remember the compile time array size check macros to enforce that then I'd happily switch to passing the format pointer directly from match data.

6by9 replied:
Actually
```
#define IMX290_NUM_FORMATS 2

static const struct imx290_pixfmt imx290_colour_formats[IMX290_NUM_FORMATS] = {
	{ MEDIA_BUS_FMT_SRGGB10_1X10, 10 },
	{ MEDIA_BUS_FMT_SRGGB12_1X12, 12 },
};

static const struct imx290_pixfmt imx290_mono_formats[IMX290_NUM_FORMATS] = {
	{ MEDIA_BUS_FMT_Y10_1X10, 10 },
	{ MEDIA_BUS_FMT_Y12_1X12, 12 },
};
```
and then using IMX290_NUM_FORMATS instead of the equivalent ARRAY_SIZE value. If someone were to ever add an extra value into the array then it'd break the compile.

Happy with that solution?

pelwell replied:
Delighted.

6by9 replied:
Done then.

6by9 committed:
media: i2c: imx290: Switch set_hmax to use imx290_write_buffered_reg

imx290_set_hmax was using two independent writes to set up hmax,
when all other multi-register writes were using imx290_write_buffered_reg
which claims the group hold first.

Switch imx290_set_hmax to using imx290_write_buffered_reg too.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

pelwell commented on drivers/media/i2c/imx290.c:
```diff
@@ -312,6 +358,11 @@ static const struct imx290_mode imx290_modes_2lanes[] = {
 		.data_size = ARRAY_SIZE(imx290_1080p_settings),
+		.clk_data = {
+			[CLK_37_125] = imx290_37_125mhz_clock_1080p,
+			[CLK_74_250] = imx290_74_250mhz_clock_1080p,
```
This should be CLK_74_25, and below. And the _4lanes modes could also do with the same treatment.

6by9 replied:
My compile was still running :-(

6by9 commented:
Branch updated based on your review comments, but my test build is still running....

pelwell commented:
Spoiler - it builds.

pelwell commented:
Happy for me to merge?

6by9 commented:
> Spoiler - it builds.

I need a faster machine, or to switch branches less.

> Happy for me to merge?

Yes. Thanks for your review.

pelwell commented:
Use git worktrees, and buildbot.
